core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+                            }
+                        }
+                        break;
+                    }
+                    tot_adv += g.adv;
+                }
+            }
+        }
+
+        overlay_color
+    }
+}
+
+/// Performs a conversion from `angle` (in degrees) to a vector (`Point`)
+/// (polar to cartesian transformation)
+/// Suitable for cases where we don't care about distance, it is assumed 1000
+///
+/// The implementation could be replaced by (cos(`angle`), sin(`angle`)),
+/// if we allow trigonometric functions.
+/// In the meantime, approximate this with predefined octagon
+fn get_vector(angle: i32) -> Point {
+    //octagon vertices
+    let v = [
+        Point::new(0, 1000),
+        Point::new(707, 707),
+        Point::new(1000, 0),
+        Point::new(707, -707),
+        Point::new(0, -1000),
+        Point::new(-707, -707),
+        Point::new(-1000, 0),
+        Point::new(-707, 707),
+    ];
+
+    let angle = angle % 360;
+    let vertices = v.len() as i32;
+    let sector_length = 360 / vertices; // only works if 360 is divisible by vertices
+    let sector = angle / sector_length;
+    let sector_angle = (angle % sector_length) as f32;
+    let v1 = v[sector as usize];
+    let v2 = v[((sector + 1) % vertices) as usize];
+    Point::lerp(v1, v2, sector_angle / sector_length as f32)
+}
+
+/// Find whether vector `v2` is clockwise to another vector v1
+/// `n_v1` is counter clockwise normal vector to v1
+/// ( if v1=(x1,y1), then the counter-clockwise normal is n_v1=(-y1,x1)
+#[inline(always)]
+fn is_clockwise_or_equal(n_v1: Point, v2: Point) -> bool {
+    let psize = v2.x * n_v1.x + v2.y * n_v1.y;
+    psize < 0
+}
+
+/// Find whether vector v2 is clockwise or equal to another vector v1
+/// `n_v1` is counter clockwise normal vector to v1
+/// ( if v1=(x1,y1), then the counter-clockwise normal is n_v1=(-y1,x1)
+#[inline(always)]
+fn is_clockwise_or_equal_inc(n_v1: Point, v2: Point) -> bool {
+    let psize = v2.x * n_v1.x + v2.y * n_v1.y;
+    psize <= 0
+}
+
+/// Draw a rounded rectangle with corner radius 2
+/// Draws only a part (sector of a corresponding circe)
+/// of the rectangle according to `show_percent` argument,
+/// and optionally draw an `icon` inside
+pub fn rect_rounded2_partial(
+    area: Rect,
+    fg_color: Color,
+    bg_color: Color,
+    show_percent: i32,
+    icon: Option<(&[u8], Color)>,
+) {
+    const MAX_ICON_SIZE: u16 = 64;
+
+    let r = area.translate(get_offset());
+    let clamped = r.clamp(constant::screen());
+
+    set_window(clamped);
+
+    let center = r.center();
+    let colortable = get_color_table(fg_color, bg_color);
+    let mut icon_colortable = colortable;
+
+    let mut use_icon = false;
+    let mut icon_area = Rect::zero();
+    let mut icon_area_clamped = Rect::zero();
+    let mut icon_data = [0_u8; ((MAX_ICON_SIZE * MAX_ICON_SIZE) / 2) as usize];
+    let mut icon_width = 0;
+
+    if let Some(i) = icon {
```

Comment:
fwiw you can write `if let Some((icon_bytes, icon_color)) = icon`

Reply:
seems nicer: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2